File: typescript/bob/bob.ts
const isUpperCase = (message: string) : boolean => message == message.toUpperCase() 
const isQuestion = (message: string) : boolean => message[message.length - 1]  == "?"
export function hey(message: string): string {
  
  message = message.trim();
  let hasLetter : boolean = /[a-zA-Z]/.test(message)

  if(isQuestion(message) && isUpperCase(message) && hasLetter)
    return "Calm down, I know what I'm doing!";
  else if(isQuestion(message))
    return "Sure.";
  else if(message == "")
    return "Fine. Be that way!";
  else if(isUpperCase(message) && hasLetter)
    return "Whoa, chill out!";
  else
    return "Whatever."

}

File: scala/bob/src/main/scala/Bob.scala
object Bob {
  def response(statement: String): String = {
    statement.trim() match {
      case s if s.isBlank() => "Fine. Be that way!"
      case s if isQuestion(s) && isYelling(s) =>
        "Calm down, I know what I'm doing!"
      case s if isQuestion(s) => "Sure."
      case s if isYelling(s)  => "Whoa, chill out!"
      case _                  => "Whatever."
    }
  }

  def isQuestion(statement: String): Boolean = statement.endsWith("?")
  def isYelling(statement: String): Boolean = {
    val statementWithOnlyAlphabets = statement.replaceAll("[^a-zA-Z]", "")
    !statementWithOnlyAlphabets
      .isBlank() && (statementWithOnlyAlphabets
      .toUpperCase() == statementWithOnlyAlphabets)
  }
}
